pkg/grpc/tls_conn_creds.go

```diff
@@ -30,13 +31,15 @@ func (c credsFromConn) ClientHandshake(_ context.Context, _ string, _ net.Conn)
 func (c credsFromConn) ServerHandshake(rawConn net.Conn) (net.Conn, credentials.AuthInfo, error) {
 	tlsConn, _ := rawConn.(interface {
 		net.Conn
-		Handshake() error
+		HandshakeContext(ctx context.Context) error
 		ConnectionState() tls.ConnectionState
 	})
 	if tlsConn == nil {
 		return rawConn, nil, nil
 	}
-	if err := tlsConn.Handshake(); err != nil {
+	ctx, cancel := context.WithTimeoutCause(context.Background(), 2*time.Second, errors.New("TLS handshake timeout"))
```

Comment:
I would add a field `tlsHandshakeTimeout` to `credsFromConn` similarly to what you did for `alpnDemuxListener`

Reply:
I think, I improved that a bit - could you check the newest version?